src/Workspaces/VisualBasic/Portable/Extensions/CastAnalyzer.vb

```diff
@@ -373,16 +382,35 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Extensions
         Private Shared Function CastRemovalChangesDefaultValue(castType As ITypeSymbol, outerType As ITypeSymbol) As Boolean
             If castType.IsNumericType Then
                 Return Not outerType.IsNumericType
-            Else
-                Select Case castType.SpecialType
-                    Case SpecialType.System_DateTime
-                        Return Not outerType.SpecialType = SpecialType.System_DateTime
-                    Case SpecialType.System_Boolean
-                        Return Not (outerType.IsNumericType OrElse outerType.SpecialType = SpecialType.System_Boolean)
-                    Case Else
-                        Return False
-                End Select
             End If
+
+            If castType.OriginalDefinition?.SpecialType = SpecialType.System_Nullable_T Then
+                ' Don't allow casts of Nothing to T? to be removed unless the outer type is T? or Object.
+                ' Otherwise, Nothing will lose its "nullness" and get the default value of T.
+                '
+                ' So, this is OK:
+                '
+                '   Dim x As Object = DirectCast(Nothing, Integer?)
+                '
+                ' But this is not:
+                '
+                '   Dim x As Integer = DirectCast(Nothing, Integer?)
+
+                If castType.Equals(outerType) OrElse outerType.SpecialType = SpecialType.System_Object Then
+                    Return False
+                End If
+
+                Return True
+            End If
+
+            Select Case castType.SpecialType
+                Case SpecialType.System_DateTime
```

Comment:
Is it really specifc types like this, or should we be looking at all value types here?